### PR TITLE
Adds initial support for external attachments using the Filesystem or MinIO

### PR DESCRIPTION
**Pull request by Spoffy**
## Context

When attachments are uploaded to Grist, they're stored in an SQLite table within the document. 

The result is that large files (multi-GB) substantially increase the size of the SQLite file, and cause performance problems (for example, with snapshots).

## Proposed solution

The current design / full scope of changes can be seen [in this Google document.](https://docs.google.com/document/d/1ST_DuR22llDyx4PVAMdBlHDz22L5QIJ4qIyLMOUOtZ8/edit#heading=h.d487w4iic6q).

This PR contains a minimal proof of concept for external attachments, with the following features:
- External attachment storage using either MinIO or the local filesystem.
- A "default attachment store" setting for documents, which sets the destination for newly uploaded attachments.
- Defaulting to existing behaviour. Without additional configuration, Grist's attachment storage behaviour is unchanged.
- MinIO attachment storage is automatically made available if it's configured for document storage, but not automatically used by documents.

No UI exists for configuring anything related to attachment storage at this time. To use a store, the API or an SQLite client needs using to set the `idOfDefaultAttachmentStore` document setting to the ID of the desired attachment store. Attachment store IDs are logged to the console.

Currently only a single store of each backend type (minio, filesystem) can exist, as they're automatically configured using the existing setting. 

## Related issues

#1021 

## Review guidance / code overview

Three significant components are added in this PR, with the rest of the PR being generally glue around them:
- `AttachmentFileManager` (in "AttachmentFileManager.ts") handles attachment fetching / saving logic on behalf of `ActiveDoc`
- An `IAttachmentStore` is an instantiated store, that handles saving / loading attachments from a specific provider.
- `AttachmentStoreProvider` is used to resolve store IDs to stores, and is the main way of providing other components (e.g AttachmentFileManager) with access to store instances.

The available types of attachment store are defined in `coreCreator.ts`, under the name "attachmentStoreBackends". 

Additionally:
- A few existing classes have been extended with streaming methods, to simplify interaction with the existing attachment code and to facilitate large files in the future without keeping copies in memory.
- Tests may not pass yet, and additional unit tests are WIP.

## Migrations / schema changes

* One migration to `_gristsys_files` has been added, which adds a column to track which store a document is in.
* `idOfDefaultAttachmentStore` has been added to the `DocumentSettings` JSON blob.

Both of these should be non-breaking on older installations, as both should be ignored. 

## Has this been tested?

Behaviour has been manually tested in the following ways:
- Filesystem storage
- MinIO client with MinIO storage
- MinIO client with S3 storage

**There are currently no unit tests**. These should be added before merging, but the goal.

### MinIO

1. Set up Grist to use MinIO normally.
2. Start Grist
3. Retrieve the attachment store id from the console.
4. Download the Grist document.
5. Open the Grist document in a database client, and add `"idOfDefaultAttachmentStore": "store id"` to `_grist_DocInfo` -> `documentSettings`
6. Upload the Grist document
7. Try adding an attachment to the new document. The name should show up in MinIO's storage path on the local filesystem.

### Known issues
- When using S3 storage, attachments showed as broken until refreshing the page.

### CODE REVIEW

File: app/server/lib/AttachmentStoreProvider.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  name: string,
+  create: (storeId: string) => IAttachmentStore,
+}
+
+interface IAttachmentStoreDetails {
+  id: string;
+  factory: IAttachmentStoreBackendFactory;
+}
+
+export class AttachmentStoreProvider implements IAttachmentStoreProvider {
```
Comment (paulfitz) [inline app/server/lib/AttachmentStoreProvider.ts:25]: Have I seen a description of how stores work somewhere in this PR? For devs to have a mental model? Here might be a good place if not?

File: app/server/lib/AttachmentStore.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    super(`Unable to create ${storeBackend} store '${storeId}'` + formattedContext);
+  }
+}
+
+// TODO - Can make this generic if *Stream methods become part of an interface.
+// TODO - Put this into another file.
+export class MinIOAttachmentStore implements IAttachmentStore {
+  constructor(
+    public id: string,
+    private _storage: MinIOExternalStorage,
```
[the remark this answers is not in the document]
Comment (paulfitz) [reply app/server/lib/AttachmentStore.ts:75]: Read rest of PR. Feels fine, no MinIO creep :+1:

**Review by paulfitz — COMMENTED** [review]
Just a batch of superficial comments from a skim. Feels like it is coming along pretty well!

File: app/common/DocumentSettings.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@ export interface DocumentSettings {
   locale: string;
   currency?: string;
   engine?: EngineCode;
+  idOfDefaultAttachmentStore?: string;
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/common/DocumentSettings.ts:5]: My motivation for phrasing it this way specifically was the ambiguity of the variable if it ends in `id`.

It could be:
`defaultAttachmentStore Id`
or
`default attachmentStoreId`

Which are semantically different interpretations of the same name! I'd be happy to swap it around if you think `defaultAttachmentStoreId` is clear enough though, despite the above :)

File: app/server/lib/AttachmentStore.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:75]: 100% - I can't remember if I left the "Make an interface for streamable external storage" in a TODO or not somewhere, but it's planned, so we can turns "MinIOAttachmentStore" into "ExternalStorageAttachmentStore"

File: app/server/lib/AttachmentStoreProvider.ts (hunk 1 above)
Comment (Spoffy) [reply app/server/lib/AttachmentStoreProvider.ts:25]: I don't think I put an overview in the code, I'll add one here :+1:

File: app/server/lib/AttachmentFileManager.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
+import { DocStorage } from "app/server/lib/DocStorage";
+import { checksumFileStream } from "app/server/lib/checksumFile";
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentFileManager.ts:2]: This is a really interesting one, as these are all IDE generated. Pretty sure my IDE here is organizing them based on numeric value, and not lowercasing them first. It's also interesting it's not flagging the quotes!

File: app/server/lib/ActiveDoc.ts [hunk 5]
```diff
@@ -2822,17 +2841,21 @@ export class ActiveDoc extends EventEmitter {
     return this._dataEngine;
   }
 
+  private async _getDocumentSettings(): Promise<DocumentSettings | undefined> {
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/ActiveDoc.ts:2844]: I can switch this to throwing an error, then wrap the code below in `_makeEngine` in a try-catch block to handle this case?

File: app/common/DocumentSettings.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulfitz) [reply app/common/DocumentSettings.ts:5]: `attachmentStoreId` is my favorite now I'm afraid

File: app/server/lib/AttachmentStoreProvider.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+export interface IAttachmentStoreProvider {
+  // Returns the store associated with the given id, returning null if no store with that id exists.
+  getStore(id: AttachmentStoreId): Promise<IAttachmentStore | null>
+
+  getAllStores(): Promise<IAttachmentStore[]>;
+
+  storeExists(id: AttachmentStoreId): Promise<boolean>;
+}
+
+export interface IAttachmentStoreBackendFactory {
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStoreProvider.ts]: I looked deep into my heart and gave it a shot, let me know if this is better!

File: app/server/lib/AttachmentFileManager.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    const store = await this._getStore(storeId);
+    if (!store) {
+      this._log.info({ fileIdent, storeId }, "tried to fetch attachment from an unavailable store");
+      throw new StoreNotAvailableError(storeId);
+    }
+    return this._addFileToAttachmentStore(store, fileIdent, fileData);
+  }
+
+  public async getFileData(fileIdent: string): Promise<Buffer | null> {
+    this._log.debug({ fileIdent }, "retrieving file data");
```
Comment (Spoffy) [inline app/server/lib/AttachmentFileManager.ts:94]: I'm concerned this logging might be a bit much. When a document is opened, it loads attachment previews. That results in 1 or 2 log lines for every attachment in a document.

File: app/server/lib/ActiveDoc.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulfitz) [reply app/server/lib/ActiveDoc.ts:2844]: that'd be fine

File: app/server/lib/AttachmentStoreProvider.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (paulfitz) [reply app/server/lib/AttachmentStoreProvider.ts]: thanks :) yes

File: app/server/lib/AttachmentFileManager.ts (hunk 7 above)
Comment (paulfitz) [reply app/server/lib/AttachmentFileManager.ts:95]: That might indeed be a bit much.

File: test/server/lib/HostedStorageManager.ts [hunk 8]
```diff
@@ -1000,15 +1004,23 @@ describe('HostedStorageManager', function() {
...
       defaultParams = [
         tmpDir,
         workerId,
         false,
         docWorkerMap,
-        {
-          setDocsMetadata: async (metadata) => {},
-          getDocFeatures: async (docId) => undefined,
-        },
+        new Callbacks(),
```
Comment (Spoffy) [inline test/server/lib/HostedStorageManager.ts:1023]: Don't think this is supposed to be in here, I think this came out of a git stash. Will remove.

File: app/server/lib/AttachmentFileManager.ts (hunk 7 above)
Comment (Spoffy) [reply app/server/lib/AttachmentFileManager.ts:112]: Removed 1 of the log lines - the other one is quite useful for debugging as it mentions the attachment pool, and which storage (document / external) the file is going into.

File: app/server/lib/AttachmentFileManager.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  public async addFile(
+    storeId: AttachmentStoreId | undefined,
+    fileExtension: string,
+    fileData: Buffer
+  ): Promise<AddFileResult> {
+    const fileIdent = await this._getFileIdentifier(fileExtension, Readable.from(fileData));
+    if (storeId === undefined) {
+      log.info(`AttachmentFileManager adding ${fileIdent} to local document storage in '${this._docName}'`);
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentFileManager.ts]: Switched all the logging to include relevant metadata (doc names, store IDs, etc).

File: app/server/lib/AttachmentStore.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts]: This is done now - any external storage should be usable, if wrapped in `ExternalStorageAttachmentStore`

File: app/server/lib/ActiveDoc.ts [hunk 10]
```diff
@@ -636,6 +648,7 @@ export class ActiveDoc extends EventEmitter {
     useExisting?: boolean,       // If set, document can be created as an overlay on
                                  // an existing sqlite file.
   }): Promise<ActiveDoc> {
+    console.log(`OPENOPENOPENOPENOPENOPENOPENOPEN ${this.docName} ANYTHING ELSE`);
```
Comment (paulfitz) [inline app/server/lib/ActiveDoc.ts:651]: Stray debugging message?

File: app/server/lib/ActiveDoc.ts [hunk 11]
```diff
@@ -388,6 +392,14 @@ export class ActiveDoc extends EventEmitter {
       loadTable: this._rawPyCall.bind(this, 'load_table'),
     });
 
+    // This will throw errors if _doc or externalAttachmentStoreProvider aren't provided, and ActiveDoc tries to use
```
Comment (paulfitz) [inline app/server/lib/ActiveDoc.ts:395]: what is _doc?

File: app/server/lib/AttachmentFileManager.ts [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+
+interface AttachmentFileManagerLogInfo {
+  fileIdent?: string;
+  storeId?: string | null;
+}
+
+/**
+ * Instantiated on a per-document to basis to provide a document with access to its attachments.
```
Comment (paulfitz) [inline app/server/lib/AttachmentFileManager.ts:70]: strike "to" in "to basis"?

File: app/server/lib/AttachmentFileManager.ts [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        fileIdent,
+        isNewFile: false,
+      };
+    }
+
+    // Possible issue if this upload fails - we have the file tracked in the document, but not available in the store.
+    // TODO - Decide if we keep an entry in SQLite after an upload error or not. Probably not?
+    await store.upload(this._getDocPoolId(), fileIdent, Readable.from(fileData));
+
+    // TODO - Confirm in doc storage that it's successfully uploaded? Need to decide how to handle a failed upload.
```
Comment (paulfitz) [inline app/server/lib/AttachmentFileManager.ts:193]: Since external storage is not entirely under our control (an admin might reasonably delete some files there) we need to degrade gracefully if some attachments are mysteriously missing from expected location. Presumably that means that even without this TODO, a failure is non-catastrophic in the sense that the document will continue to work.

File: app/server/lib/AttachmentStore.ts [hunk 14]
```diff
@@ -0,0 +1,174 @@
+import { joinKeySegments, StreamingExternalStorage } from "app/server/lib/ExternalStorage";
+import * as fse from "fs-extra";
+import * as stream from "node:stream";
+import * as path from "path";
+
+export type DocPoolId = string;
+type FileId = string;
+
+
+// Minimum required info from a document
```
Comment (paulfitz) [inline app/server/lib/AttachmentStore.ts:10]: "Minimum required info from a document" for what purpose?

File: app/server/lib/AttachmentStore.ts [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+ * Gets the correct pool id for a given document, given the document's id and trunk id.
+ *
+ * Attachments are stored in a "Document Pool", which is used to manage the attachments' lifecycle.
+ * Document pools are shared between snapshots and forks, but not between documents. This provides quick forking
+ * and snapshotting (not having to copy attachments), while avoiding more complex systems like reference tracking.
+ *
+ * Generally, the pool id of a document should be its trunk id if available (because it's a fork),
+ * or the document's id (if it isn't a fork).
+ *
+ * This means that attachments need copying to a new pool when a document is copied.
```
Comment (paulfitz) [inline app/server/lib/AttachmentStore.ts:29]: I don't think we've talked explicitly about the consequences of document copying now being a potentially slow operation. As long as a copy fails cleanly if there's a problem, I think that would be the main thing.

File: app/server/lib/AttachmentStoreProvider.ts [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+export class AttachmentStoreProvider implements IAttachmentStoreProvider {
+  private _storeDetailsById: { [storeId: string]: IAttachmentStoreDetails } = {};
+
+  constructor(
+    _backends: IAttachmentStoreSpecification[],
+    _installationUuid: string
+  ) {
+    // In the current setup, we automatically generate store IDs based on the installation ID.
+    // The installation ID is guaranteed to be unique, and we only allow one store of each backend type.
```
Comment (paulfitz) [inline app/server/lib/AttachmentStoreProvider.ts:46]: `we only allow one store of each backend type` - I thought we agreed to move on from this as too restrictive? It makes no sense to me to allow say an s3 bucket and an azure bucket but then to kick up a fuss if the azure bucket is replaced with a distinct s3 bucket.

Comment (paulfitz) [reply app/server/lib/AttachmentStoreProvider.ts:46]: I remember discussing permitting a small label that would take precedence over storeSpec.name if given.

**Review by paulfitz — COMMENTED** [review]
Is copying of documents with external attachments addressed? Might have missed it. If not, would want to be sure we don't miss that in follow-up.

File: app/server/lib/ActiveDoc.ts (hunk 10 above)
Comment (Spoffy) [reply app/server/lib/ActiveDoc.ts:651]: Yup, this and others. Thanks for catching this :sweat_smile:

File: app/server/lib/AttachmentFileManager.ts (hunk 13 above)
Comment (Spoffy) [reply app/server/lib/AttachmentFileManager.ts:193]: It's not well explained in this PR, but the transfer code defines this error handling more explicitly.

In the case that the file isn't accessible in external storage, it should show that the attachment isn't available in the UI.  I think that's handled on the clientside when the API returns an error response?

File: app/server/lib/AttachmentStore.ts (hunk 15 above)
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:29]: I don't think we have either - might be worth a chat? 

I think the copy behaviour is missing from this PR to be honest (:grimacing:). It might be worth opening as a separate PR at this point?

File: app/server/lib/AttachmentStoreProvider.ts (hunk 16 above)
Comment (Spoffy) [reply app/server/lib/AttachmentStoreProvider.ts:46]: We've since discussed this, and will use a "JSON in environment variable" approach

File: app/server/lib/AttachmentStore.ts (hunk 15 above)
Comment (paulfitz) [reply app/server/lib/AttachmentStore.ts:29]: Made a kanban and added this to it https://grist.quip.com/zDbRAal05oe8/External-Attachments-TOC#temp:C:fPF3b8cf6575b3d49a7bd5d05efd.

File: app/server/lib/ActiveDoc.ts (hunk 10 above)
Comment (paulfitz) [reply app/server/lib/ActiveDoc.ts:651]: Do you plan on removing it last minute? Just checking why it is still present.

File: app/server/lib/AttachmentFileManager.ts (hunk 13 above)
Comment (paulfitz) [reply app/server/lib/AttachmentFileManager.ts:193]: Moved your question to kanban

File: app/server/lib/AttachmentStoreProvider.ts (hunk 16 above)
Comment (paulfitz) [reply app/server/lib/AttachmentStoreProvider.ts:46]: For the purposes of this PR, what do you plan to do about the `we only allow one store of each backend type` restriction currently written here?

File: test/server/lib/ActiveDoc.ts [hunk 17]
```diff
@@ -23,6 +23,7 @@ import { createDocTools } from 'test/server/docTools';
 import * as testUtils from 'test/server/testUtils';
 import { EnvironmentSnapshot } from 'test/server/testUtils';
 import * as tmp from 'tmp';
+import { AttachmentStoreProvider } from "../../../app/server/lib/AttachmentStoreProvider";
```
Comment (jordigh) [inline test/server/lib/ActiveDoc.ts:26]: Should be an absolute import, right?

File: app/server/utils/MemoryWritableStream.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import { Writable } from "stream";
+
+// Creates a writable stream that can be retrieved as a buffer.
+// Sub-optimal implementation, as we end up with *at least* two copies in memory one in `buffers`, and one
```
Comment (jordigh) [inline app/server/utils/MemoryWritableStream.ts:4]: What are the performance implications of this? Are we going to be having memory problems with moderately large files?

File: app/server/lib/coreCreator.ts [hunk 19]
```diff
@@ -13,6 +18,24 @@ export const makeCoreCreator = () => makeSimpleCreator({
       create: configureMinIOExternalStorage,
     },
   ],
+  attachmentStoreOptions: [
+    {
+      name: 'minio',
+      isAvailable: async () => checkMinIOExternalStorage() !== undefined,
+      // TODO - Move this to another file
```
Comment (jordigh) [inline app/server/lib/coreCreator.ts:25]: What do you envision should be done here? Which file?

**Review by jordigh — COMMENTED** [review]
It's a fairly complex setup, but given the problem, I think you have convinced me that all of this complexity is required. I could not find many opportunities for simplification.

File: app/server/lib/AttachmentStoreProvider.ts (hunk 16 above)
Comment (Spoffy) [reply app/server/lib/AttachmentStoreProvider.ts:46]: As discussed in Slack, I'll put together a separate PR with all the config logic that will allow this. :slightly_smiling_face:

Comment (paulfitz) [reply app/server/lib/AttachmentStoreProvider.ts:46]: Ok, I added it to the kanban. Resolved for the purposes of this PR.

File: app/server/lib/ActiveDoc.ts (hunk 10 above)
Comment (Spoffy) [reply app/server/lib/ActiveDoc.ts]: Nah, likely an oversight on my part when doing the other bits of your feedback. Glad this wasn't ticked off at least!

File: app/server/lib/AttachmentStore.ts [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+
+export type DocPoolId = string;
+type FileId = string;
+
+
+// Minimum required info from a document
+// Compatible with Document entity for ease of use
+export interface AttachmentStoreDocInfo {
+  id: string;
+  // This should NOT be an optional: the programmer should make an explicit choice for this value to prevent
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:14]: It's inherently a nullable value - a document might not have trunkId. 

However, if I make it an optional (i.e typescripts `?` syntax), code like this won't error:
```javascript
getDocPoolIdFromDocInfo({ id: _someDoc.id })
```
The above is incorrect - and would result in attachments ending up in the wrong place (the fork's attachment file pool, instead of the original document's file pool).

This is correct:
```javascript
getDocPoolIdFromDocInfo({ id: _someDoc.id, trunkId: _someDoc.trunkId })
```

So by omitting `?` and using `string | null | undefined` instead, it's hopefully catching a class of errors early.

File: app/server/lib/AttachmentStore.ts [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+  // accidental omission breaking attachments.
+  trunkId: string | null | undefined;
+}
+
+/**
+ * Gets the correct pool id for a given document, given the document's id and trunk id.
+ *
+ * Attachments are stored in a "Document Pool", which is used to manage the attachments' lifecycle.
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:22]: 120 is what we've got configured in .eslint.js. Happy to go lower personally, but is it worth discussing at a Grist level? We could lower the max length for comments only.

File: app/server/lib/AttachmentStore.ts [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+  }
+}
+
+export class ExternalStorageAttachmentStore implements IAttachmentStore {
+  constructor(
+    public id: string,
+    private _storage: StreamingExternalStorage,
+    private _prefixParts: string[]
+  ) {
+    if (_storage.removeAllWithPrefix == undefined) {
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:102]: Yup! Honestly, a simple `!_storage.removeAllWithPrefix` would do. `removeAllWithPrefix` *needs* to be a valid function for this to work.

File: app/server/lib/AttachmentStore.ts [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+      throw new InvalidAttachmentExternalStorageError("ExternalStorage does not support removeAllWithPrefix");
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:103]: It's a new method on the interface, so would cause compilation errors (e.g in SaaS) when this was merged.

This was the approach I agreed with Paul instead to prevent that. 

(I.e - Not all external storage implementers will be valid to use with this adapter class when this is merged).

File: app/server/lib/MinIOExternalStorage.ts [hunk 24]
```diff
@@ -182,21 +189,44 @@ export class MinIOExternalStorage implements ExternalStorage {
...
+      bucketItemStream
+        .on('error', reject)
+        .on('end', () => {
+          resolve(results);
+        })
+        .on('data', data => {
+          results.push(data);
+        });
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/MinIOExternalStorage.ts:231]: Oooh, this is interesting. Didn't know there was an async iterator approach to stream reading.

I can't see any reason this would break / change the semantics of the function - so yeah, this is great :)

File: test/server/lib/ActiveDoc.ts (hunk 17 above)
Comment (Spoffy) [reply test/server/lib/ActiveDoc.ts:26]: :pray:

File: app/server/utils/MemoryWritableStream.ts (hunk 18 above)
Comment (Spoffy) [reply app/server/utils/MemoryWritableStream.ts:4]: It's potentially one more copy kept in memory than we do currently? Although I'm not 100% clear on how many copies we keep currently (at least one, but not sure SQLite handles reads here).

There might be some room to optimise this by allocating progressively larger buffers, and providing an option to allocate a specific initial buffer size. Might be tricky to get right?

Ideally though I'd want to use that effort to move us fully to streaming, so we're not keeping any copies of files in memory. That's out of scope of the current changes though, as a lot of the existing code passes buffers around.

For performance impact - I'd guess it depends on how big your files are, and how much memory you have.
If you're uploading 10GB files on a 4GB RAM server, it's definitely going to have a bad day (although that's no different to currently anyway). I'd probably say 1/4 of server memory is a safe value? But that's a bit of a guess.

File: app/server/lib/coreCreator.ts (hunk 19 above)
Comment (Spoffy) [reply app/server/lib/coreCreator.ts:25]: ....Honestly not sure. At the time I wrote this comment, it was likely we'd have a `MinIOAttachmentStore` file. But that got generalised into `ExternalStorageAttachmentStore`.

 It feels long for where it is. Might be just worth moving it to the end of this file and referencing it, to keep the coreCreator definition easy to read. 
 
 What do you think?

File: app/server/lib/AttachmentStore.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (jordigh) [reply app/server/lib/AttachmentStore.ts:14]: Oh, boy, "optional" vs "an optional", subtle reading thing there.

Maybe rephrase, "We explicitly make this a union type instead of making the attribute optional because the programmer must make a conscious choice to mark it as null or undefined, not merely omit it."

File: app/server/lib/AttachmentStore.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (jordigh) [reply app/server/lib/AttachmentStore.ts:102]: I think so. I kind of consider the `==` operator to be "legacy". At least checking for falsey is clearer in intent.

File: app/server/lib/AttachmentStore.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (jordigh) [reply app/server/lib/AttachmentStore.ts:103]: I see, makes sense.

File: app/server/utils/MemoryWritableStream.ts (hunk 18 above)
Comment (jordigh) [reply app/server/utils/MemoryWritableStream.ts:4]: If it's not worse than it was, it seems good enough. :+1:

File: app/server/lib/coreCreator.ts (hunk 19 above)
Comment (jordigh) [reply app/server/lib/coreCreator.ts:25]: It's pretty short, but if you want to move it to the bottom of the file, that's fine too.

It would also be fine to get rid of the TODO comment. Actually, that's what I really want, to reduce the number of TODOs by one.

File: app/server/lib/AttachmentStore.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (jordigh) [reply app/server/lib/AttachmentStore.ts:23]: It won't be much of a problem beyond diffs. If you can easily warp comments at a lower character count, nice. If you can't, don't worry about it and let's move on.

[the remark this answers is not in the document]
Comment (Spoffy) [reply app/server/lib/AttachmentStore.ts:23]: Updated IDE with settings for this for the future, probably a bit too big to go through and update all comments right now :)

**Review by jordigh — APPROVED** [review]

File: app/server/lib/ExternalStorage.ts [hunk 25]
```diff
@@ -5,6 +5,7 @@ import {createTmpDir} from 'app/server/lib/uploads';
 import {delay} from 'bluebird';
 import * as fse from 'fs-extra';
 import * as path from 'path';
+import stream from "node:stream";
```
Comment (paulfitz) [inline app/server/lib/ExternalStorage.ts:8]: lint mumble mumble

File: app/server/lib/FlexServer.ts [hunk 26]
```diff
@@ -27,11 +27,16 @@ import {createSandbox} from 'app/server/lib/ActiveDoc';
 import {attachAppEndpoint} from 'app/server/lib/AppEndpoint';
 import {appSettings} from 'app/server/lib/AppSettings';
 import {attachEarlyEndpoints} from 'app/server/lib/attachEarlyEndpoints';
+import {
+  AttachmentStoreProvider, checkAvailabilityAttachmentStoreOptions, IAttachmentStoreProvider
+} from "app/server/lib/AttachmentStoreProvider";
```
Comment (paulfitz) [inline app/server/lib/FlexServer.ts:32]: lint here and elsewhere

File: app/server/lib/ICreate.ts [hunk 27]
```diff
@@ -15,10 +15,11 @@ import {createSandbox, SpawnFn} from 'app/server/lib/NSandbox';
 import {SqliteVariant} from 'app/server/lib/SqliteCommon';
 import {ITelemetry} from 'app/server/lib/Telemetry';
 import {IDocStorageManager} from './IDocStorageManager';
-import { Comm } from "./Comm";
-import { IDocWorkerMap } from "./DocWorkerMap";
-import { HostedStorageManager, HostedStorageOptions } from "./HostedStorageManager";
-import { DocStorageManager } from "./DocStorageManager";
+import {Comm} from "./Comm";
```
Comment (paulfitz) [inline app/server/lib/ICreate.ts:18]: lint lint lints

File: app/server/lib/MinIOExternalStorage.ts [hunk 28]
```diff
@@ -182,21 +189,38 @@ export class MinIOExternalStorage implements ExternalStorage {
...
     // Stick to this maximum in case we are using this client to talk to AWS.
     const N = this._batchSize || 1000;
-    for (let i = 0; i < versions.length; i += N) {
-      const iVersions = versions.slice(i, i + N).filter(v => v) as string[];
-      if (iVersions.length === 0) { continue; }
-      await this._s3.removeObjects(this.bucket, iVersions.map(versionId => {
-        return { name: key, versionId };
-      }));
+    for (let i = 0; i < objects.length; i += N) {
+      const batch = objects.slice(i, i + N);
```
Comment (paulfitz) [inline app/server/lib/MinIOExternalStorage.ts:212]: what was the `filter(v => v)` in the old code?

File: test/server/docTools.ts [hunk 29]
```diff
@@ -17,6 +17,7 @@ import {tmpdir} from 'os';
 import * as path from 'path';
 import * as tmp from 'tmp';
 import {create} from "app/server/lib/create";
+import { AttachmentStoreProvider, IAttachmentStoreProvider } from "../../app/server/lib/AttachmentStoreProvider";
```
Comment (paulfitz) [inline test/server/docTools.ts:20]: avoid ../../ in imports in grist-core please (I know there's a few)

File: test/server/lib/ActiveDoc.ts [hunk 30]
```diff
@@ -6,6 +6,7 @@ import * as gristTypes from 'app/common/gristTypes';
 import { GristObjCode } from 'app/plugin/GristData';
 import { TableData } from 'app/common/TableData';
 import { ActiveDoc } from 'app/server/lib/ActiveDoc';
+import { AttachmentStoreProvider } from "app/server/lib/AttachmentStoreProvider";
```
Comment (paulfitz) [inline test/server/lib/ActiveDoc.ts:9]: linty lint

File: test/server/lib/FilesystemAttachmentStore.ts [hunk 31]
```diff
@@ -0,0 +1,73 @@
+import { FilesystemAttachmentStore } from "app/server/lib/AttachmentStore";
+import { MemoryWritableStream } from "app/server/utils/MemoryWritableStream";
+import { assert } from "chai";
+import { createTmpDir } from "../docTools";
+import { mkdtemp, pathExists } from 'fs-extra';
```
Comment (paulfitz) [inline test/server/lib/FilesystemAttachmentStore.ts:5]: Please use internally consistent import quotes in few files

Why am I the import quote police

File: test/server/lib/HostedStorageManager.ts [hunk 32]
```diff
@@ -4,6 +4,7 @@ import {SCHEMA_VERSION} from 'app/common/schema';
 import {DocWorkerMap, getDocWorkerMap} from 'app/gen-server/lib/DocWorkerMap';
 import {HomeDBManager} from 'app/gen-server/lib/homedb/HomeDBManager';
 import {ActiveDoc} from 'app/server/lib/ActiveDoc';
+import { AttachmentStoreProvider, IAttachmentStoreProvider } from "app/server/lib/AttachmentStoreProvider";
```
Comment (paulfitz) [inline test/server/lib/HostedStorageManager.ts:7]: lint

File: test/upgradeDocumentImpl.ts [hunk 33]
```diff
@@ -4,15 +4,59 @@
  * Usage:
  *    test/upgradeDocument <docPaths...>
  */
-import {copyFile} from 'app/server/lib/docUtils';
-import {createDocTools} from 'test/server/docTools';
+import { DocStorage } from "app/server/lib/DocStorage";
+import { DocStorageManager } from "app/server/lib/DocStorageManager";
+import { copyFile } from 'app/server/lib/docUtils';
```
Comment (paulfitz) [inline test/upgradeDocumentImpl.ts:9]: Lint on some of these

**Review by paulfitz — COMMENTED** [review]
Some minor comments, overall looks good, want to do some testing before approving.

File: app/server/lib/MinIOExternalStorage.ts (hunk 28 above)
Comment (Spoffy) [reply app/server/lib/MinIOExternalStorage.ts:212]: Not entirely sure - looks to be filtering out empty strings and undefined, but not sure which call site is passing those values down to here.

That check still exists anyway - all code which used to use `_deleteBatch`  uses `_deleteVersions` above, which does `filter(v => v)` before passing the values down to `_deleteObjects`.

`_deleteObjects` is only used by `removeAllWithPrefix` directly, which passes it the output from S3's `_listObjects`, so should all be valid?

**Review by paulfitz — APPROVED** [review]
Thanks @Spoffy, this looks like a good start!

**Comment by vviers** [comment]
Following up on this, the correct key to set isn't `idOfDefaultAttachmentStore` but `attachmentStoreId`
